### PR TITLE
Fix #9930: "change" replaces 0-param projections by constants

### LINKED ISSUE #9930: `change tt with tt` is not identity
#### Description of the problem

```coq
Set Primitive Projections.
Record params := { width : nat }.
Definition p : params := Build_params 64.

Goal width p = 0 -> width p = 0.
  intros.
  let e := lazymatch type of H with ?e = 0 => e end in
  change tt with tt in H;
  let E := lazymatch type of H with ?E = 0 => E end in
  idtac "before:" e; idtac "after :" E;
  (* before: (width p) *)
  (* after : (width p) *)
  tryif constr_eq e E then admit else idtac.
  all : fail "not equal".
  (* Error: Tactic failure: not equal. *)
Abort.
```

cc @samuelgruetter 

Note fixing this makes #9512 need a new test.

See also #9929

#### Coq Version

8.9.0

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fixes #9930.

### CODE REVIEW

**Comment by RalfJung** [comment]
I think to actually close the issue when the PR lands you need the "Fix #xx" in the PR text, not the title.

Also, the issue says
> Note fixing this makes #9512 need a new test.

Is that still true?

**Comment by SkySkimmer** [comment]
>I think to actually close the issue when the PR lands you need the "Fix #xx" in the PR text, not the title.

In any commit message also works.

File: pretyping/tacred.ml [hunk 1]
```diff
@@ -1009,11 +1009,11 @@ let change_map_constr_with_binders_left_to_right g f (env, l as acc) sigma c =
     let app = (mkApp (hdf, Array.sub al 0 (Array.length al - 1))) in
     let app' = f acc app in
     let a' = f acc a in
-      (match EConstr.kind sigma app' with
-      | App (hdf', al') when hdf' == hdf ->
-        (* Still the same projection, we ignore the change in parameters *)
-        mkProj (p, a')
-      | _ -> mkApp (app', [| a' |]))
+    let hdf', _ = decompose_app_vect sigma app' in
+    if hdf' == hdf then
```
Comment (ppedrot) [inline pretyping/tacred.ml]: I know you did not introduce this, but the use of physical equality looks very fragile to me. Can't you use a more robust heuristic?

Comment (ppedrot) [reply pretyping/tacred.ml]: In particular, `hdf` is guaranteed to be a `Const` node, you could take advantage of that.

Comment (SkySkimmer) [reply pretyping/tacred.ml]: I don't really want to mess with it.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Will merge soon.

**Comment by gares** [comment]
I never understood why change tt with tt was actually doing something ;-)

**Pull request merged**